### PR TITLE
Remove the forward class hint feature.

### LINKED ISSUE #5613: Forward hints :>> are not documented.
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5613
From: @Zimmi48
Reported version: 8.6

### PR DESCRIPTION
**Pull request by ppedrot**
It was not documented, not properly tested and thus likely buggy. Judging from the code alone I spotted already one potential bug. Further more it was prominently making use of the infamous "arbitrary term as hint" feature.

Since the only user in our CI seems to be a math-classes file that introduced the feature under a claim of "cleanup", I believe we can safely remove it without anyone noticing.

EDIT: no more overlays needed.

Closes #5613

### CODE REVIEW

**Comment by JasonGross** [comment]
What were these hints supposed to do?

**Comment by ppedrot** [comment]
They were declaring additional hints defined as specialized variants of the projections for their subclasses, it seems. It was an automatic way to declare subinstances thus.

**Comment by JasonGross** [comment]
I thought `:>` already did that when the thing being defined was a class?

**Comment by ppedrot** [comment]
Summoning @mattam82

File: test-suite/output/UnivBinders.out
Comment (SkySkimmer) [inline test-suite/output/UnivBinders.out:72]: What's causing this change?

File: pretyping/typeclasses.mli
Comment (SkySkimmer) [inline pretyping/typeclasses.mli:16]: What's this used for?

File: test-suite/output/UnivBinders.out
Comment (ppedrot) [reply test-suite/output/UnivBinders.out:72]: Probably one less retyping that used to generate fresh universes.

File: pretyping/typeclasses.mli
Comment (ppedrot) [reply pretyping/typeclasses.mli:16]: Nothing, I could remove it but then it's more invasive API-wise.

Comment (SkySkimmer) [reply pretyping/typeclasses.mli:16]: Invade on imo

Comment (ppedrot) [reply pretyping/typeclasses.mli]: Done. I took the occasion to introduce a record type for methods.

**Comment by gares** [comment]
overlay merged

**Comment by ppedrot** [comment]
@SkySkimmer do you want to assign?

File: vernac/vernacexpr.ml
Comment (SkySkimmer) [inline vernac/vernacexpr.ml:110]: Can we make this a real type instead of `unit option`?

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot: merge now

File: vernac/vernacexpr.ml
Comment (SkySkimmer) [reply vernac/vernacexpr.ml:110]: https://github.com/coq/coq/pull/13128

**Comment by Zimmi48** [comment]
Hum, sorry for having failed to review the code of this PR before, but you should not remove features without a changelog entry, even when the feature was undocumented and largely unused. Even the removal of `Chapter` (which was not used at all in our CI) did get a changelog entry.

Furthermore, is the feature that was removed the `:>>` appearing [here](https://coq.github.io/doc/master/refman/language/core/assumptions.html#grammar-token-of-type). If yes, great, that's less to document, but the syntax appearing in the manual should then have been adjusted as well.

**Comment by Zimmi48** [comment]
I'm also slightly puzzled that we did not wait for an answer from @mattam82 before merging this PR. Did one of the author or the assignee talk to him physically?

**Comment by ppedrot** [comment]
> Did one of the author or the assignee talk to him physically?

Yep, @mattam82 confirmed via private communication that this could be safely removed (it was an experiment IIUC).

**Comment by ppedrot** [comment]
> Did one of the author or the assignee talk to him physically?

Yep, @mattam82 confirmed via private communication that this could be safely removed (it was an experiment IIUC).

**Comment by Zimmi48** [comment]
OK, good to hear. I'll prepare the PR with the changelog / documentation update since you didn't react to my other comment.

**Pull request merged**
